### ISSUE TITLE
Support Preseed File Templating

### ISSUE THREAD

**Issue by timblaktu** [hashicorp/packer#10605]
Consider the user scenario presented in #10583, in which 
1. A packer template is designed to build one source that includes, say, a virtualbox-iso builder. 
2. The build block invokes this source multiple times, with the only difference being the preseed file that is used by the builder. 
3. The preseed files for the variants are identical except for a few spots, e.g. package selection and late_command.
4. Therefore from a maintenance perspective, it is highly desirable to dynamically generate the preseed files from a template, instead of maintaining multiple nearly identical text files in SCM.

Before I implemented this using a preparatory jinja2 template/rendering process before calling packer, I thought I'd check here if there was an existing way to do this.

This feature request assumes there is not, and proposes an addition to packer that leverages its internal templating mechanism to support preseed file templating.

**Comment by timblaktu** [comment]
@azr would you have a recommendation for the best way to do conditionally include a build source in an HCL2 packer build block?

Here I went ahead and implemented preseed file generation with a python jinja2 template process that precedes the packer call in my Jenkinsfile. The `preseed_variants` come from Jenkins job extendedChoice parameter that takes on the string values: [headless, gnome, mate, xfce, kde, lxde, cinnamon].  

```
...
                                # Render the preseed files from template for each variant packer is going to build below
                                pipenv run python3 - << EOF
from jinja2 import Environment, FileSystemLoader
env = Environment(loader=FileSystemLoader(searchpath="./"))
assert ',' not in ${params.base_images.toString()}, "extendedChoice parameter should not be inserting commas!"
for preseed_variant in ${params.base_images.toString()}.split(' '):
    print("Generating preseed file for {}...".format(preseed_variant))
    with open("./http/preseed-{}.cfg".format(preseed_variant), 'w') as fh:
        fh.write(env.get_template('preseed.cfg.j2').render(preseed_variant=preseed_variant))
EOF
...
```

The resulting preseed files are referenced from the boot_commands defined in each variant's build source block. (See below)

This is working fine for the below packer build block, which uses hard-coded sources, for ALL variants. But, since the variants are actually a variable (parameter) in the Jenkins job, I'd like to make the inclusion of the packer sources conditional. 

```
build {                                                                                                                                                                                                                                                         name = "debian10"
  description = <<EOF
This build creates Debian 10 base images for the following builders:
* virtualbox-iso
It builds separate headless and gnome-desktop versions in parallel
and places them in uniquely-namemd output folders.
This build is intended to run on an ansible-managed Jenkins agent.
EOF
  source "source.virtualbox-iso.debian10" {
    name                    = "headless"
    vm_name                 = "{{build_name}}-headless-${local.scm_suffix}"
    output_directory        = "${var.output_directory_root}/{{build_name}}-headless-${local.scm_suffix}"
    boot_command            = concat(local.boot_command_start, local.boot_command_preseed_url_headless, local.boot_command_end)
  }
  source "source.virtualbox-iso.debian10" {
    name                    = "gnome"
    vm_name                 = "{{build_name}}-gnome-${local.scm_suffix}"
    output_directory        = "${var.output_directory_root}/{{build_name}}-gnome-${local.scm_suffix}"
    boot_command            = concat(local.boot_command_start, local.boot_command_preseed_url_gnome, local.boot_command_end)
  }
  source "source.virtualbox-iso.debian10" {
    name                    = "mate"
    vm_name                 = "{{build_name}}-mate-${local.scm_suffix}"
    output_directory        = "${var.output_directory_root}/{{build_name}}-mate-${local.scm_suffix}"
    boot_command            = concat(local.boot_command_start, local.boot_command_preseed_url_mate, local.boot_command_end)                                                                                                                                     }
  source "source.virtualbox-iso.debian10" {                                                                                                                                                                                                                       name                    = "xfce"
    vm_name                 = "{{build_name}}-xfce-${local.scm_suffix}"
    output_directory        = "${var.output_directory_root}/{{build_name}}-xfce-${local.scm_suffix}"                                                                                                                                                              boot_command            = concat(local.boot_command_start, local.boot_command_preseed_url_xfce, local.boot_command_end)                                                                                                                                     }
  source "source.virtualbox-iso.debian10" {
    name                    = "kde"
    vm_name                 = "{{build_name}}-kde-${local.scm_suffix}"
    output_directory        = "${var.output_directory_root}/{{build_name}}-kde-${local.scm_suffix}"
    boot_command            = concat(local.boot_command_start, local.boot_command_preseed_url_kde, local.boot_command_end)
  }
  source "source.virtualbox-iso.debian10" {
    name                    = "lxde"
    vm_name                 = "{{build_name}}-lxde-${local.scm_suffix}"
    output_directory        = "${var.output_directory_root}/{{build_name}}-lxde-${local.scm_suffix}"
    boot_command            = concat(local.boot_command_start, local.boot_command_preseed_url_lxde, local.boot_command_end)
  }
  source "source.virtualbox-iso.debian10" {
    name                    = "cinnamon"
    vm_name                 = "{{build_name}}-cinnamon-${local.scm_suffix}"
    output_directory        = "${var.output_directory_root}/{{build_name}}-cinnamon-${local.scm_suffix}"
    boot_command            = concat(local.boot_command_start, local.boot_command_preseed_url_cinnamon, local.boot_command_end)
  }
  provisioners ...
```

**Comment by azr** [comment]
Hey @timblaktu, I think this would be an important feature and wanted to find a way to address that one; after some research, my current thinking is that it would be just awesome to have something like the `templatefile` function from Terraform: https://www.terraform.io/docs/language/functions/templatefile.html

I think that we should think about the possibility to enforce filename extensions though; so that editors can highlight these files.

**Comment by timblaktu** [comment]
Thanks @azr. I realize now that _Some_ of what I'm asking here can be achieved by simply using the existing support for `-only` cli arg. My Jenkinsfile defines job parameters for the build "variants" (headless, gnome, kde, ...), as a list of strings. Instead of having the packer sources be defined conditionally, IMO now it maybe makes more since to statically define them all up front, then "only" run the ones specified in the jenkins build params. This is easily done in the Jenkinsfile when it calls `packer build`.

But the part about generating preseed files is the primary aspect of this issue so I'll leave it open, and I agree with you, it'd be great to be able to do arbitrary templating in packer, instead of having to pre-process using python/jinja2.

**Comment by ekristen** [comment]
having `templatefile` from terraform would be amazing. right now I have a nice packer template setup for people to re-use but at the moment you have to do a bunch of search and replace to change like the default user in auto-unattend because there's no variable interpolation.

**Comment by sylviamoss** [comment]
Closing this one. The `templatefile` was added in https://github.com/hashicorp/packer/pull/10776 and will be released with 1.7.1 🎉
